controllers/edgedeployment_controller.go

```diff
@@ -64,71 +66,160 @@ func (r *EdgeDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+	for _, edgeDevice := range edgeDevices {
+		err := r.removeDeploymentFromDevice(ctx, edgeDeployment.Name, edgeDevice)
+		if err != nil {
+			return err
+		}
+	}
+	return r.EdgeDeploymentRepository.RemoveFinalizer(ctx, edgeDeployment, YggdrasilDeviceReferenceFinalizer)
+
 }
 
-func newEdgeDeviceOwnerReference(ed *managementv1alpha1.EdgeDevice) metav1.OwnerReference {
-	blockOwnerDeletion := true
-	isController := false
-	return metav1.OwnerReference{
-		APIVersion:         ed.GroupVersionKind().GroupVersion().String(),
-		Kind:               ed.GetObjectKind().GroupVersionKind().Kind,
-		Name:               ed.GetName(),
-		UID:                ed.GetUID(),
-		BlockOwnerDeletion: &blockOwnerDeletion,
-		Controller:         &isController,
+func (r *EdgeDeploymentReconciler) removeDeploymentFromDevice(ctx context.Context, name string, edgeDevice managementv1alpha1.EdgeDevice) error {
+	var newDeployments []managementv1alpha1.Deployment
+	for _, deployment := range edgeDevice.Status.Deployments {
+		if deployment.Name != name {
+			newDeployments = append(newDeployments, deployment)
+		}
+	}
+	edgeDevice.Status.Deployments = newDeployments
+	err := r.EdgeDeviceRepository.UpdateStatus(ctx, &edgeDevice)
+	if err != nil {
+		return err
+	}
+
+	deviceCopy := edgeDevice.DeepCopy()
+	if deviceCopy.Labels != nil {
+		delete(deviceCopy.Labels, name)
+		err = r.EdgeDeviceRepository.Patch(ctx, &edgeDevice, deviceCopy)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
-func (r *EdgeDeploymentReconciler) addOwnerReference(ctx context.Context, owner *managementv1alpha1.EdgeDevice, ownedDeployment *managementv1alpha1.EdgeDeployment) error {
-	ownerRefs := ownedDeployment.GetOwnerReferences()
-	if ownerRefs == nil {
-		ownerRefs = []metav1.OwnerReference{}
-	}
-	if hasOwnerReference(ownerRefs, owner) {
-		return nil
+func (r *EdgeDeploymentReconciler) removeDeploymentFromNonMatchingDevices(ctx context.Context, name string, matchingDevices, labelledDevices []managementv1alpha1.EdgeDevice) error {
+	matchingDevicesMap := make(map[string]struct{})
+	for _, device := range matchingDevices {
+		matchingDevicesMap[device.Name] = struct{}{}
 	}
-	ownerReference := newEdgeDeviceOwnerReference(owner)
 
-	ownerRefs = append(ownerRefs, ownerReference)
+	for _, device := range labelledDevices {
+		if _, ok := matchingDevicesMap[device.Name]; !ok {
+			err := r.removeDeploymentFromDevice(ctx, name, device)
+			if err != nil {
+				return err
+			}
+		}
+	}
 
-	newEdgeDeployment := ownedDeployment.DeepCopy()
-	newEdgeDeployment.SetOwnerReferences(ownerRefs)
+	return nil
+}
 
-	return r.EdgeDeploymentRepository.Patch(ctx, ownedDeployment, newEdgeDeployment)
+func (r *EdgeDeploymentReconciler) addDeploymentsToDevices(ctx context.Context, name string, edgeDevices []managementv1alpha1.EdgeDevice) error {
+outer:
+	for _, edgeDevice := range edgeDevices {
```

Comment:
running this one in parallel makes sense since we expect a lot of edge devices and minimize a failure of a single device to prevent updating the others.

Reply:
I'd like to postpone that optimization until we see performance issues.

Reply:
I've created https://issues.redhat.com/browse/ECOPROJECT-221 to track the optimization-related changes